src/app-manager.ts

```diff
@@ -1,37 +1,70 @@
 import { PluginProfile, ModuleProfile, API } from './types'
 import { Plugin } from './plugin'
 
+interface EventListeners {
+  [origin: string]: {
+    [type: string]: {
+      [key: string]: (value: any) => void
+    }
+  }
+}
+
 export class AppManager {
-  events = {}
+  private modules: {
+    [type: string]: { json: ModuleProfile, api: API }
+  } = {}
+  private plugins: {
+    [type: string]: { json: PluginProfile, api: Plugin }
+  } = {}
+  events: EventListeners = {}
 
   constructor(dependancies: {
     modules?: { json: ModuleProfile; api: API }[]
     plugins?: { json: PluginProfile; api: Plugin }[]
+    options?: {
+      boostrap: string
+    }
   }) {
     // Modules
     const modules = dependancies.modules || []
-    modules.forEach(({ json, api }) => this.addApi(json, api))
+    this.modules = modules.reduce((acc, {json, api}) => {
+      this[api.type] = {}
+      this.activateApi(json, api) // Activate module automaticaly
+      return { ...acc, [json.type]: { json, api } }
+    }, {})
 
     // Plugins
     const plugins = dependancies.plugins || []
-    plugins.forEach(({ json, api }) => {
-      this.addApi(json, api)
+    this.plugins = plugins.reduce((acc, {json, api}) => {
+      this[api.type] = {}
+      return { ...acc, [json.type]: { json, api } }
+    }, {})
+
+    // bootstrap
+    if (dependancies.options && dependancies.options.boostrap) {
+      const {api} = this.modules[dependancies.options.boostrap]
+      api['activate'].on((type: string) => this.activate(type))
+      api['deactivate'].on((type: string) => this.deactivate(type))
+    }
+  }
 
-      api.request = ({ type, key, value }) => this[type][key](value)
 
-      const notifications = json.notifications || []
-      notifications.forEach(({ type, key }) => {
-        if (!this.events[api.type]) this.events[api.type] = {}
-        if (!this.events[api.type][type]) this.events[api.type][type] = {}
-        this.events[api.type][type][key] = api.notifs[type][key]
-      })
-    })
+  /** Broadcast a message to every plugin listening */
+  private broadcast(type: string, key: string, value: any) {
+    for (const origin in this.events) {
+      if (this.events[origin][type]) {

```

Comment:
here the root key is `origin` ( in `this.events[origin][type]` ) .
but line 91, `api.type` is used...

Reply:
Line 91, the origin of the event triggered is the `api.type`, but I can create a variable to make it easier to understand. Something like: `const origin = api.type`.

Reply:
At least we should make obvious in the code that `api.type` is an URL (is it ?). at first I thought that `api.type` is the reference name of the plugin.

Reply:
I guess for module it is not an URL, but there should be some token describing the location of the module.

Reply:
It's not an url (it has nothing to do with event.origin). It's the type of the plugin listening on this event;
Basically if "ethdoc" listen on the method "compilationFinished" of module "solCompiler" we would have : 
- origin : "ethdoc"
- type: "solCompiler"
- key: "compilationFinished"
What word would you use for that?

Reply:
Sure. What about `listener`?

Reply:
sounds good!